### PR TITLE
Allow Navigation Through Subdirectories

### PR DESCRIPTION
**Pull request by omarnyte**
~~## FileDirectoryHandler
I've created this new `Handler` to replace the custom handler passed in to the `Router` inside of `Main`. Like all `Handler`'s, this class exposes one method: `generateResponse(Request request)`. It uses the `DataStore`'s new `#isDirectory(String uri)` method to determine whether it should generate a response from the `FileHandler` or the `DirectoryHandler`.~~

~~My concern with this method is that pretty much any request to a URI that does not exist will go to the `FileHandler` since the `Request`s are routed based on whether or not they are directories, and a nonexistent directory is not a directory. Using the word "routed" in the previous sentence tells me that `FileDirectoryHandler` is doing something other than what it's name suggests (and that the logic should probably go inside `Router`). I'm only now realizing this, but I'm going to leave this up so that you two are aware of my though process.~~

## Router 
As I realized above, it doesn't make a ton of sense to route `Request`s to a `FileHandler` as a default when the `Request`'s URI might be for a directory. This really should be the responsibility of the `Router`, so I've moved the logic there. This required updating the `Main` class in order to pass a `DataStore` to the `Router`. 

## NotFoundHandler
I've returned to using a `NotFoundHandler` to build 404 `Response`s as a default `Handler`. 

## RouterTest
I've known for a while that I need to decouple actual implementations of `DataStore`/`Directory` from different unit tests. I'm finally trying that with a static `MockDirectory` inside of this class. While it's still simple to test that `Request`s are routed to the correct custom `Handler` or the default `Handler`, I'm not entirely sure how to test that `Requests` are properly routed to the `DirectoryHandler` or the `FileHandler` without relying on the implementation of those `Handler`s. `Router` only exposes the `#getResponse(Request request)` method. Because neither `DirectoryHandler` nor `FileHandler` are passed in as arguments, I can't think of a non-invasive way to mock their `Response`s. I could potentially integrate setters to configure default `Handler`s for files and directories, but it also seems strange to me given that those defaults will likely not change. 

```Java
//Router.java
  public Response getResponse(Request request) {
    Handler handler = getHandler(request.getURI());
    return handler.generateResponse(request);
  }

  private Handler getHandler(String uri) {
    try {
      if (this.routes.get(uri) != null) {
        return this.routes.get(uri);
      } else if (this.store.isDirectory(uri)) {
        DataStore subdirectoryStore = this.store.createSubdirectoryStore(uri);
        return new DirectoryHandler(subdirectoryStore, uri);
      } else if (this.store.isFile(uri)) {
        return new FileHandler(this.store);
      }
    } catch(NonexistentDirectoryException e) {
      System.err.println(e);
    }

    return defaultHandler;
  }
```

## RootHandler --> DirectoryHandler
Given that I reused the RootHandler in order to create the `messageBody` for subdirectories, I renamed `RootHandler` to `DirectoryHandler`. 

#### `#createSubdirectoryStore(String uri)` 
In order to reuse the `RootHandler`/`DirectoryHandler` for subdirectories, I added the `#createSubdirectoryStore(String uri)` interface method to `DataStore`. The method is very simple: 
```Java
  public Directory createSubdirectoryStore(String uri) throws NonexistentDirectoryException {
    return new Directory(this.directoryPath + uri);
  }
```
My two concerns are:
1. How do I test the method? The method doesn't physically add a new directory inside the data store. Instead, it's an abstraction for an existing subdirectory. I don't want to test that a subdirectory was created using any of the other methods inside of `Directory`. The only other option that I can think about is to assert that the created subdirectory is an instance of the `Directory` class, but I know that that isn't a robust test.  
2. Creating a `new Directory` potentially `throws` a `NonexistentDirectoryException`. This exception was added because a user could potentially pass an invalid directory path as a CLI argument when starting the server. When creating a subdirectory, though, that probably won't happen because the subdirectories are being created only when the `Request`'s URI hits a directory. Thus, throwing the `NonexistentDirectoryException` from the interface method couples any other future`DataStore` types to an exception specific to a `Directory`. 

#### `#postFile(String uri, byte[] content)` 
Within the same classes, I also want to improve my error catching in the `#postFile(String uri, byte[] content)` method. Right now, an `IOException` within `Directory`s implementation of the method won't stop the `FormHandler` from building a `Response` with a random file name, even if that file was not created because on an `IOException`. It's not hard to see why: 

```Java
//FormHandler.java
  private Response buildPostResponse(Request request) {
      String uri = "/POSTed/" + randomFileName() + ".txt";
      byte[] content = createFileContent(request);
      store.postFile(uri, content);
      return new Response.Builder(HttpStatusCode.SEE_OTHER)
                         .setHeader("Location", uri)
                         .build();
  }
```
```Java
// Directory.java
  public void postFile(String uri, byte[] content) {
    try {
      File file = new File(this.directoryPath + uri);
      file.createNewFile();
      FileOutputStream ouputStream = new FileOutputStream(file.getAbsolutePath());
      ouputStream.write(content);
    } catch(IOException e) {
      System.err.println("IOException failed inside Directory");
      System.err.println(e);
    }
  }
```
If I threw (rather than `catch`) the `IOException` from `Directory`'s implementation of `#postFile` the `DataStore` interface method would also have to `throws` the  exception. This would couple the interface method to an implementation using a `Directory`. It's basically the same problem I have with `#createSubdirectoryStore`. 

 My idea was to create a custom exception (perhaps `InternalServerError`) and catch it inside any `Handler` that calls `#postFile`. The catch could `build` and return a 500 `Response`. I haven't had a lot of exposure to custom exceptions outside of this project, so I'm not sure if I'm going overboard with the custom exceptions. So far, I have four exceptions (`BadRequestException`, `MissingFlagException`, `NonexistentDirectoryException`, and `UnsupportedFlagException`). What should I be considering when deciding to create a custom exception vs. relying on an existing exception? Is there a way that I can better catch the `IOException` inside of `#postFile`? The `File` class returns a boolean for `.createNewFile()` or `mkdir()`. I can do the same with `#postFile`, but I don't know if there are any drawbacks to using a boolean when a method has a side effect like creating a file. 

On a slightly related note, I'm thinking that I should change the name of `#postFile` to something like `#addFileToStore`. `POST` is an HTTP concept, and adding a file to a directory or a database isn't really tied to a network.

### CODE REVIEW

File: src/main/java/DataStore/DataStore.java [hunk 1]
```diff
@@ -1,9 +1,11 @@
 public interface DataStore {
 
+  public Boolean isDirectory(String uri);
+  public Boolean isFile(String uri);
   public String[] listContent();
   public Boolean existsInStore(String uri);
   public byte[] readFile(String uri);
   public String getFileType(String uri);
   public void postFile (String uri, byte[] content);
```
Comment (jayden) [inline src/main/java/DataStore/DataStore.java]: The name `postFile` is a bit unclear. Is it creating a file?

File: src/main/java/Router.java [hunk 2]
```diff
@@ -3,10 +3,12 @@
...
+  private DataStore store;
 
-  public Router(Handler defaultHandler, HashMap<String, Handler> routes) {
+  public Router(Handler defaultHandler, HashMap<String, Handler> routes, DataStore store) {
     this.defaultHandler = defaultHandler;
     this.routes = routes;
+    this.store = store;
   }
 
   public Router(HashMap<String, Handler> routes) {
```
Comment (jayden) [inline src/main/java/Router.java]: Is this constructor still being used?

File: src/main/java/Handler/FileDirectoryHandler.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
+public class FileDirectoryHandler implements Handler {
```
Comment (jayden) [inline src/main/java/Handler/FileDirectoryHandler.java]: What's the difference between this and DirectoryHandler?

File: src/main/java/DataStore/DataStore.java (hunk 1 above)
Comment (omarnyte) [reply src/main/java/DataStore/DataStore.java]: It is creating a file, so I've changed the name to `#createFileWithContent`.

**Pull request merged**